### PR TITLE
feature: upload and instantiation of multisig contracts

### PR DESCRIPTION
**Pull request by reductionista**
## Uploading

For using terra multisig, first upload the wasm files for these contracts:

yarn gauntlet upload cw4_group
yarn gauntlet upload cw3_flex_multisig

## Create a cw4 group

This instantiates a group with 3 members, whose votes are weighted by 25, 25, and 50 respectively:  (passing admin address is optional)

```
yarn gauntlet cw4_group:deploy --members='[terra1pl4k5rj2jv6phm2vvhkttju7px6va2ja2v3haw,terra1tsxn3zzp09kvwpx03gzwquhc6nn794vvznuhzr,terra1s66cck3sxacdc2jfpdd4t4pk4yzc60pa72ssdr]' --admin=terra1pl4k5rj2jv6phm2vvhkttju7px6va2ja2v3haw --weights='[25,25,50]' --network=local
```
## Create a cw3 flex multisig wallet

This instantiates a multisig wallet, for the group above... with a max voting period of 28800s (8 hours) and an absolute threshold of 50 votes:

```
yarn gauntlet cw3_flex_multisig:deploy --network=local --group_addr=terra179sumv8jjahdp6w6w7sze4n0rc9jsf4u5wpd25 --max_voting_period='{ "time" : 28800}' --threshold='{ "absolute_count" : { "weight" : 50 }}'
```

You may also specify the threshold as a percentage of the total votes (the same, in this case) or even a quorum combined with a threshold.
( See https://docs.cosmwasm.com/cw-plus/0.9.0/cw3/cw3-flex-spec for details )

## Multisig execution commands

After the group & multisig wallet have been created, we can support the following operations (TODO):

- `gauntlet CONTRACT:COMMAND:propose` - propose a new transaction
- `gauntlet CONTRACT:COMMAND:vote` - cast a vote on a proposal
- `gauntlet CONTRACT:COMMAND:execute` - execute a proposal which has gained enough votes
- `gauntlet CONTRACT:COMMAND:close` - close a proposal, so no more votes are allowed

### CODE REVIEW

File: packages-ts/gauntlet-terra-contracts/src/commands/contracts/multisig/create.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+  if (flags.input) return flags.input as WalletInitParams
+  return {
+    group_addr: flags.group_addr,
+    max_voting_period: flags.max_voting_period,
+    threshold: flags.threshold
+  }
+}
+
+const validateWalletInitParams = (params: WalletInitParams): boolean => {
+   let group_addr: string = params.group_addr;
```
Comment (RodrigoAD) [inline packages-ts/gauntlet-terra-contracts/src/commands/contracts/multisig/create.ts:59]: There's no need of reassigning this, you can use `params` directly

File: packages-ts/gauntlet-terra-contracts/deps/contracts/cw3_flex_multisig/schema/instantiate_msg.json [hunk 2]
```diff
@@ -0,0 +1,135 @@
+{
```
[the remark this answers is not in the document]
Comment (reductionista) [reply packages-ts/gauntlet-terra-contracts/deps/contracts/cw3_flex_multisig/schema/instantiate_msg.json:1]: I'm concerned about keeping the wasm files and schemas in sync.  The schemas are generated from the wasm files and must match.  If we remove one, seems like we should remove both.  But then we'd need to make sure we copy them into the right folder whenever we check out our repo, before running any multisig commands.

This repo depends on 3 external artifacts: `cw20_base.wasm`, `cw3_flex_multisig.wasm`, and `cw4_group.wasm`.  All of those come from `https://github.com/CosmWasm/cw-plus/releases`.  It looks like we do store a copy of `cw20_base.wasm` in this repo here:
https://github.com/smartcontractkit/chainlink-terra/tree/main/tests/e2e/common_artifacts

Can we just move the other two wasm files related to it there?  At some point, we'll need end-to-end tests that make use of them too.

[the remark this answers is not in the document]
Comment (reductionista) [reply packages-ts/gauntlet-terra-contracts/deps/contracts/cw3_flex_multisig/schema/instantiate_msg.json:1]: Actually, I guess a better solution is to have them dynamically downloaded, in the same way `gauntlet upload` does for our own artifacts.  Currently, that command will only work for these with the `version=local` option... I'll see if I can modify it to be able to pull them from `cw-plus` repo... I think that might solve this problem.

File: cw3_flex_multisig_deploy.log [hunk 3]
```diff
@@ -0,0 +1,12 @@
+[dominos-mbp:~/workspace/chainlink-terra> yarn gauntlet  cw3_flex_multisig:deploy --network=local --group_addr=terra1js4j4838l3yvp7n32nqjkyvvepj0f25melmemf --max_voting_period='{ "time" : 28800}' --threshold='{ "absolute_count" : { "weight" : 50 }}'
```
Comment (RodrigoAD) [inline cw3_flex_multisig_deploy.log:1]: Is this log here for some reason? Let's remove it otherwise

File: cw4_group_deploy.log [hunk 4]
```diff
@@ -0,0 +1,48 @@
+Bombay-12:
+
+[dominos-mbp:~/workspace/chainlink-terra> yarn gauntlet cw4_group:deploy --members='[terra1pl4k5rj2jv6phm2vvhkttju7px6va2ja2v3haw,terra1tsxn3zzp09kvwpx03gzwquhc6nn794vvznuhzr,terra1s66cck3sxacdc2jfpdd4t4pk4yzc60pa72ssdr]' --admin=terra1pl4k5rj2jv6phm2vvhkttju7px6va2ja2v3haw --weights='[34,33,33]' --network=bombay-testnet
```
Comment (RodrigoAD) [inline cw4_group_deploy.log:3]: Same with this one

File: packages-ts/gauntlet-terra-contracts/deps/contracts/cw3_flex_multisig/schema/instantiate_msg.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (RodrigoAD) [reply packages-ts/gauntlet-terra-contracts/deps/contracts/cw3_flex_multisig/schema/instantiate_msg.json:1]: That seems like the best solution.

File: packages-ts/gauntlet-terra-contracts/setBillngs.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
+import { BN } from '@chainlink/gauntlet-core/dist/utils'
```
Comment (RodrigoAD) [inline packages-ts/gauntlet-terra-contracts/setBillngs.ts:1]: I think this was included accidentally? Seems that the file name was changed

File: packages-ts/gauntlet-terra-contracts/src/commands/contracts/multisig/wallet.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
+import { CATEGORIES } from '../../../lib/constants'
+import { AbstractInstruction, instructionToCommand } from '../../abstract/wrapper'
+import { Duration, isValidDuration, Threshold, isValidThreshold } from './lib/types'
+import { isValidAddress } from '../../../lib/schema'
+
+type WalletInitParams = {
```
Comment (RodrigoAD) [inline packages-ts/gauntlet-terra-contracts/src/commands/contracts/multisig/wallet.ts:6]: Let's clearly differentiate the command input vs the contract input in every command for consistency. Command input should always be on camelCase.
`WalletInitParams` is the contract input, but not necessarily the command

File: packages-ts/gauntlet-terra-contracts/deps/contracts/cw3_flex_multisig/schema/instantiate_msg.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (RodrigoAD) [reply packages-ts/gauntlet-terra-contracts/deps/contracts/cw3_flex_multisig/schema/instantiate_msg.json:1]: If we finally can't go with that, let's just move everything to `/artifacts/` , we'll look for a solution to this later

File: packages-ts/gauntlet-terra-contracts/setBillngs.ts (hunk 5 above)
Comment (RodrigoAD) [reply packages-ts/gauntlet-terra-contracts/setBillngs.ts:1]: Please remove this file

File: packages-ts/gauntlet-terra-contracts/src/commands/contracts/multisig/group.ts [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  if (input.admin && !isValidAddress(input.admin)) throw new Error('Admin is not valid')
+  return true
+}
+
+const makeContractInput = async (input: CommandInput): Promise<ContractInput> => {
+  return {
+    members: input.owners.map((owner) => ({
+      addr: owner,
+      // Same weight for every owner
+      weight: 1,
```
Comment (RodrigoAD) [inline packages-ts/gauntlet-terra-contracts/src/commands/contracts/multisig/group.ts:37]: @reductionista Please refactor this to make every owner to have the same weight

**Review by RodrigoAD — APPROVED** [review]

**Review by RodrigoAD — DISMISSED** [review]

**Review by RodrigoAD — APPROVED** [review]

**Pull request merged**